### ISSUE TITLE
ENH: Implement all np.*(x) functions as np.ndarray.*(x)

### ISSUE THREAD

**Issue by YodaEmbedding** [numpy/numpy#24081]
### Proposed new feature or change:

All single-argument `np.FUNC` functions can be implemented as `np.ndarray.FUNC`.

**Example usage:**

```python
# Currently invalid. :(
x.ceil()
```

**Rationale:**

- [`np.ndarray`](https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray) already implements some methods:

```python
all
any
argmax
argmin
argpartition
argsort
astype
byteswap
choose
clip
compress
conj
conjugate
copy
cumprod
cumsum
diagonal
dump
dumps
fill
flatten
getfield
item
itemset
max
mean
min
newbyteorder
nonzero
partition
prod
ptp
put
ravel
repeat
reshape
resize
round
searchsorted
setfield
setflags
sort
squeeze
std
sum
swapaxes
take
tobytes
tofile
tolist
tostring
trace
transpose
var
view
```

- It is unclear to a user which ones have been implemented and which ones haven't unless they try one and see if their code works/fails. (Or google the docs for each one they want to use.) For instance, `np.ndarray.ceil` does not exist, but `np.ndarray.round` does.

- Fully implementing this API style improves interchangeability with libraries such as PyTorch. With this change, there is less need to implement different methods (e.g. `torch.sin` or `np.sin`) for functions that just want to think about numerical computations on tensors of any type (`np.ndarray` or `torch.Tensor`), e.g.:

```python
def sin2(x: np.ndarray | torch.Tensor):
    return (x ** 2).sin()
```

Instead of:

```python
def sin2(x: np.ndarray | torch.Tensor):
    if isinstance(x, np.ndarray):
        return np.sin(x ** 2)
    else:
        return (x ** 2).sin()
```

- Fluent API styles are very natural for chained computations.

```python
((x ** 2).mean(axis=-1).sqrt() > x.mean(axis=-1).sqrt()).nonzero()
```

versus:

```python
np.nonzero(np.sqrt(np.mean(x ** 2, axis=-1)) > np.sqrt(np.mean(x, axis=1)))
```

---

A (not necessarily exhaustive) list of possible single-argument FUNCs:

```python
negative
positive
absolute
fabs
rint
sign
conj
conjugate
exp
exp2
log
log2
log10
expm1
log1p
sqrt
square
cbrt
reciprocal
sin
cos
tan
arcsin
arccos
arctan
sinh
cosh
tanh
arcsinh
arccosh
arctanh
degrees
radians
deg2rad
rad2deg
invert
logical_not
isfinite
isinf
isnan
isnat
fabs
signbit
spacing
modf
frexp
floor
ceil
trunc
```

**Comment by rkern** [comment]
It is generally the opinion of the numpy devs that there are too many methods on `ndarray` and that many of the existing methods should have been left as functions. We will not be adding any more, except those that are added to the [Array API standard](https://data-apis.org/array-api/2022.12/) for interoperability purposes.

**Comment by davidgilbertson** [comment]
Oh that's a pity, I just came to the GitHub issues to search for exactly this, wanting to contribute moving some methods across.

I'd be curious to know how "too many" is quantified. I use NumPy, Pandas and PyTorch quite a lot and when using PyTorch and Pandas I have never thought that there were "too many" methods. Perhaps you're talking about too many from a NumPy maintainability perspective, not from a user experience perspective, in which case I have no opinion and trust your judgement. But it's a pity that user experience suffers as a result. How do the PyTorch folks cope?

If you _are_ talking about "too many" from a user's perspective though, I can tell you this is my number one pet peeve with NumPy, it just seems so random and illogical to be able to do `arr.mean()` but not `arr.median()`.

**Comment by davidgilbertson** [comment]
Thanks for the elaboration. That all seems reasonable enough, although I'd say there's no cognitive load for the user if either ALL or NONE of the functions are available as methods. It's only the SOME case that puts a strain on my brain. And I suspect that there aren't many users out there who would be fine with NumPy removing the `mean()` method because it's available as a function already, so the case for NONE seems odd to me (if purely thinking of UX, not maintenance).

You mentioned `.argpartition()`, which I haven't used, but can tell you a true story that just the other day I typed `arr.flatnonzero()` only to be reminded that this is not a thing I'm allowed to do and I needed to do `np.flatnonzero(arr)` instead.

But they're all user problems, that still leaves the fact that it's a maintenance headache, so fair enough.

It's a shame there isn't a low-overhead way to define the logic once and have it available as either a top-level function or a method, something that works for maintainers _and_ users.
